src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs

```diff
@@ -32,5 +37,129 @@ public SafeFileHandle(IntPtr preexistingHandle, bool ownsHandle) : base(ownsHand
 
         protected override bool ReleaseHandle() =>
             Interop.Kernel32.CloseHandle(handle);
+
+        internal static unsafe SafeFileHandle Open(string fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, long preallocationSize)
+        {
+            using (DisableMediaInsertionPrompt.Create())
+            {
+                Debug.Assert(fullPath != null);
+
+                if (FileStreamHelpers.ShouldPreallocate(preallocationSize, access, mode))
+                {
+                    IntPtr fileHandle = NtCreateFile(fullPath, mode, access, share, options, preallocationSize);
+
+                    return ValidateFileHandle(new SafeFileHandle(fileHandle, ownsHandle: true), fullPath, (options & FileOptions.Asynchronous) != 0);
```

Comment:
There's a bunch of other options being set below, e.g. disabling Inheritable, setting SECURITY_SQOS_PRESENT and SECURITY_ANONYMOUS, etc... why is that not relevant if a preallocation size is set?

Reply:
It's set, but in a different place: https://github.com/dotnet/runtime/blob/d5cef92384ae43a503340417009c9b3c5713241a/src/libraries/Common/src/Interop/Windows/NtDll/Interop.NtCreateFile.cs#L81-L87

I decided to follow @JeremyKuhne suggestion and stop using `CreateFileW` and `NtCreateFile` and use `NtCreateFile` only. There is a minor perf gain (2% on opening the file) but the code became much cleaner

Reply:
How are we handling https://docs.microsoft.com/en-us/windows/win32/devnotes/calling-internal-apis ?

Reply:
@stephentoub for some reason `NtCreateFile()` has two docs:

* https://docs.microsoft.com/en-us/windows/win32/api/winternl/nf-winternl-ntcreatefile (says that it's internal)
* https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-ntcreatefile (no mention of internal)

I am not sure which doc is right, but we have been using `NtCreateFile` for more than 3 years now: https://github.com/dotnet/corefx/pull/27195

Reply:
Does NtCreateFile support every kind of file the higher level CreateFileW does?